docs/proposals/THV-2106-virtual-mcp-server.md

```diff
@@ -0,0 +1,1035 @@
+- **service_account**: Use stored credentials for backend
+- **header_injection**: Add authentication headers from secrets
+- **mapped_claims**: Transform JWT claims for backend requirements
+
+Example authentication flow:
+```
+Client → (Bearer token for Virtual MCP) → Virtual MCP
+         ├→ GitHub backend (token exchanged for GitHub PAT)
+         ├→ Jira backend (original token passed through)
+         └→ Internal DB (service account token injected)
+```
+
+#### 5. Request Routing
+
+Routes MCP protocol requests to appropriate backends:
+- Tool calls routed based on tool-to-workload mapping
+- Resource requests routed by resource namespace
+- Prompt requests handled similarly
+- Load balancing for duplicate capabilities
+
+#### 6. Elicitation Support in Composite Tools
+
+Composite tools can include elicitation steps to request additional information from users during workflow execution, following the [MCP elicitation specification](https://modelcontextprotocol.io/specification/2025-06-18/client/elicitation). This enables:
+
+- **Interactive workflows**: Request user input between steps
+- **Safety confirmations**: Require approval before destructive operations
+- **Dynamic parameters**: Gather context-dependent information
+- **Conditional execution**: Branch workflow based on user responses
+
+Elicitation steps use JSON Schema to define the structure of requested data (limited to flat objects with primitive properties). The Virtual MCP server forwards elicitation requests to the client and captures responses in three forms:
+
+- **accept**: User provided data (accessible via `{{.steps.step_id.content}}`)
+- **decline**: User explicitly rejected the request
+- **cancel**: User dismissed without choosing
+
+Subsequent steps can reference elicitation results through template expansion and use the `condition` field to execute conditionally based on user responses. The `on_decline` and `on_cancel` handlers control workflow behavior for non-acceptance scenarios.
+
+---
+
+## Authentication Architecture
+
+Virtual MCP implements a **unified configuration model with platform-specific adapters**. The same mental model applies whether running via CLI or Kubernetes, with different serialization formats optimized for each platform.
+
+### Two Authentication Boundaries
+
+Virtual MCP operates with two distinct, independent authentication boundaries:
+
+1. **Incoming Authentication** (Client → Virtual MCP)
+   - **Purpose**: Validates client requests to Virtual MCP
+   - **Token Audience**: `aud=vmcp`
+   - **Configuration**: `incomingAuth` (K8s) / `incoming_auth` (CLI)
+   - **Implementation**: Uses existing ToolHive auth middleware (OIDC, local, anonymous)
+
+2. **Outgoing Authentication** (Virtual MCP → Backends)
+   - **Purpose**: Per-backend token exchange for accessing backend services
+   - **Token Audience**: Backend-specific (e.g., `aud=github-api`, `aud=jira-api`)
```

Comment:
I'm confused when you say `github-api` do you mean a token issued by github's IDP meant to access github's API or a token issued by the same IDP as the vMCP meant to access the back end MCP server?